groups/bsl/bslstl/bslstl_ratio.t.cpp

```diff
@@ -0,0 +1,383 @@
+            ASSERT(            1000000ULL == bsl::mega::num);
+            ASSERT(         1000000000ULL == bsl::giga::num);
+            ASSERT(      1000000000000ULL == bsl::tera::num);
+            ASSERT(   1000000000000000ULL == bsl::peta::num);
+            ASSERT(1000000000000000000ULL == bsl::exa::num);
+
+        #ifdef BSL_RATIO_SUPPORTS_EXTENDED_SI_TYPEDEFS
+            ASSERT(1 == bsl::zepto::num);
+            ASSERT(1 == bsl::yocto::num);
+            ASSERT(   1000000000000000000000ULL == bsl::zepto::den);
+            ASSERT(1000000000000000000000000ULL == bsl::yocto::den);
+
+            ASSERT(1 == std::zetta::den);
+            ASSERT(1 == std::yotta::den);
+            ASSERT(   1000000000000000000000ULL == bsl::zetta::num);
+            ASSERT(1000000000000000000000000ULL == bsl::yotta::num);
+        #endif
+        }
+
+        if (verbose) printf("\tTest ratio comparisons.\n");
+        {
+            typedef bsl::ratio<3, 2> ThreeHalfs;
+            typedef bsl::ratio<6, 4> Six4ths;
+            typedef bsl::ratio<0, 1> Zero;
+            typedef bsl::ratio<-2, 3> NegTwo3rds;
+
+            ASSERT( (bsl::ratio_less<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT( (bsl::ratio_less<NegTwo3rds, Zero>::value));
+            ASSERT(!(bsl::ratio_less<ThreeHalfs,Zero>::value));
+            ASSERT(!(bsl::ratio_less<ThreeHalfs, Six4ths>::value));
+
+            ASSERT( (bsl::ratio_less_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT( (bsl::ratio_less_equal<NegTwo3rds, Zero>::value));
+            ASSERT(!(bsl::ratio_less_equal<ThreeHalfs,Zero>::value));
+            ASSERT( (bsl::ratio_less_equal<ThreeHalfs, Six4ths>::value));
+
+            ASSERT(!(bsl::ratio_greater<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT(!(bsl::ratio_greater<NegTwo3rds, Zero>::value));
+            ASSERT( (bsl::ratio_greater<ThreeHalfs,Zero>::value));
+            ASSERT(!(bsl::ratio_greater<ThreeHalfs, Six4ths>::value));
+
+            ASSERT(!(bsl::ratio_greater_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT(!(bsl::ratio_greater_equal<NegTwo3rds, Zero>::value));
+            ASSERT( (bsl::ratio_greater_equal<ThreeHalfs,Zero>::value));
+            ASSERT( (bsl::ratio_greater_equal<ThreeHalfs, Six4ths>::value));
+
+            ASSERT(!(bsl::ratio_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT(!(bsl::ratio_equal<NegTwo3rds, Zero>::value));
+            ASSERT(!(bsl::ratio_equal<ThreeHalfs,Zero>::value));
+            ASSERT( (bsl::ratio_equal<ThreeHalfs, Six4ths>::value));
+
+            ASSERT( (bsl::ratio_not_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT( (bsl::ratio_not_equal<NegTwo3rds, Zero>::value));
+            ASSERT( (bsl::ratio_not_equal<ThreeHalfs,Zero>::value));
+            ASSERT(!(bsl::ratio_not_equal<ThreeHalfs, Six4ths>::value));
+
+            // same tests, with using the _v suffix
+            ASSERT( (bsl::ratio_less_v<NegTwo3rds, ThreeHalfs>));
```

Comment:
We need to guard all the `foo_v` testing with:
```
#ifdef BSLS_LIBRARYFEATURES_HAS_CPP14_BASELINE_LIBRARY
```

Reply:
I don't think that's correct. I think we need to check if either `BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY` is defined (in which case the underlying STL provides the `_v`s) or `BSLS_LIBRARYFEATURES_HAS_CPP14_BASELINE_LIBRARY` is defined (in which case the bslstl_ratio header defines them)